server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
+            if (clusterBlocks != null) {
+                for (ClusterBlock clusterBlock : clusterBlocks) {
+                    if (clusterBlock.id() == block.block.id()) {
+                        // Reuse the existing UUID-based block
+                        indexBlock = clusterBlock;
+                        break;
+                    }
+                }
+            }
+            if (indexBlock == null) {
+                // Create a new UUID-based block
+                indexBlock = createUUIDBasedBlock(block.block);
+            }
+            assert Strings.hasLength(indexBlock.uuid()) : "Block should have a UUID";
+            blocks.addIndexBlock(index.getName(), indexBlock);
+            blockedIndices.put(index, indexBlock);
+            // update index settings as well to match the block
+            final IndexMetadata indexMetadata = metadata.getSafe(index);
+            if (block.setting().get(indexMetadata.getSettings()) == false) {
+                final Settings updatedSettings = Settings.builder()
+                    .put(indexMetadata.getSettings()).put(block.settingName(), true).build();
+
+                metadata.put(IndexMetadata.builder(indexMetadata)
+                    .settings(updatedSettings)
+                    .settingsVersion(indexMetadata.getSettingsVersion() + 1));
+            }
+        }
+
+        logger.info(() -> new ParameterizedMessage("adding block {} to indices {}", block.name,
+            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.joining(","))));
+        return ClusterState.builder(currentState).blocks(blocks).metadata(metadata).routingTable(routingTable.build()).build();
+    }
+
+    public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request,
+                              ActionListener<AddIndexBlockResponse> listener) {
+        final Index[] concreteIndices = request.indices();
+        if (concreteIndices == null || concreteIndices.length == 0) {
+            throw new IllegalArgumentException("Index name is required");
+        }
+        List<String> writeIndices = new ArrayList<>();
+        SortedMap<String, IndexAbstraction> lookup = clusterService.state().metadata().getIndicesLookup();
+        for (Index index : concreteIndices) {
+            IndexAbstraction ia = lookup.get(index.getName());
+            if (ia != null && ia.getParentDataStream() != null && ia.getParentDataStream().getWriteIndex().getIndex().equals(index)) {
+                writeIndices.add(index.getName());
+            }
+        }
+        if (writeIndices.size() > 0) {
+            throw new IllegalArgumentException("cannot add a block to the following data stream write indices [" +
+                Strings.collectionToCommaDelimitedString(writeIndices) + "]");
+        }
+
+        clusterService.submitStateUpdateTask("add-index-block " + Arrays.toString(concreteIndices),
```

Comment:
Super minor, but perhaps
```suggestion
        clusterService.submitStateUpdateTask("add-index-block-[" + request.getBlock() + "]-" Arrays.toString(concreteIndices),
```

So we can see the intended block in the pending tasks list